File: pubsub_cpp/src/subscriber.cpp
#include "zenoh.hxx"
#include <vector>
#include <string>
#include <thread>
#include <chrono>
#include <iostream>


using namespace zenoh;

int main(int argc, char** argv) {
    Config config = Config::create_default();
    auto session = Session::open(std::move(config));

    std::vector<uint8_t> received_payload;

    auto subscriber = session.declare_subscriber(
        KeyExpr("demo/example/bytes"),
        [&received_payload](const Sample& sample) {
            received_payload = sample.get_payload().as_vector();
        },
        closures::none
    );

    std::cout << "Subscriber ready." << std::endl;

    while(true){
        if (!received_payload.empty()) {
            std::cout << "Received message: "
                      << reinterpret_cast<const char*>(received_payload.data())
                      << std::endl;
            received_payload.clear();
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    

    return 0;
}

File: pubsub_py/subscriber.py
import time
import zenoh

def listener(sample):
    payload = sample.payload.to_bytes()
    print(f"Received on {sample.key_expr}: {payload}")

def main():
    session = zenoh.open(zenoh.Config())
    sub = session.declare_subscriber("vr/1/states", listener)
    print("Subscriber running. Press Ctrl+C to exit.")
    while True:
        time.sleep(1)

if __name__ == "__main__":
    main()
